### PR TITLE
Debtor debt

### PR DESCRIPTION
**Pull request by lomamech**
Aged Debtor
- Adding API for Aged Debtors
- Resolve conflict with the branch Master	
- Adding test integration for AgedDebtors			
- Adding test end to end			

@sfount

### CODE REVIEW

File: client/src/partials/reports/modals/agedDebtors.config.js [hunk 1]
```diff
@@ -0,0 +1,58 @@
+'use strict';
+
```
Comment (mbayopanda) [inline client/src/partials/reports/modals/agedDebtors.config.js]: You can remove `use strict`, gulp inject it automatically

File: client/src/partials/reports/modals/agedDebtors.config.js [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+// dependencies injection
+AgedDebtorsConfigController.$inject = [ '$state', '$http', '$uibModalInstance', 'CashboxService', 'NotifyService', 'LanguageService' ];
+
+/**
+ * AgedDebtors config controller
+ *
+ * @description
+ * This controller is responsible of cash flow report, that report include
+ * all incomes minus all depenses
```
Comment (mbayopanda) [inline client/src/partials/reports/modals/agedDebtors.config.js:12]: This description is not correct for an `Aged Report for clients`

File: server/config/routes.js [hunk 3]
```diff
@@ -551,6 +552,4 @@ exports.configure = function configure(app) {
 
   // @todo - this should use the JSON renderer instead of it's own route!
   app.get('/finance/cashflow', financeReports.cashflow.report);
-
-  //app.get('/reports/saved/finance/debtors/aged');
```
Comment (mbayopanda) [inline server/config/routes.js]: 👍

File: server/controllers/finance/reports/agedDebtors/index.js [hunk 4]
```diff
@@ -0,0 +1,209 @@
...
+ * @module finance/aged Debtors
+ *
+ * @requires lodash
+ * @requires node-uuid
+ * @requires moment
+ * @requires lib/db
+ * @requires lib/util
+ * @requires lib/ReportManager
+ * @requires lib/errors/BadRequest
+ */
```
Comment (mbayopanda) [inline server/controllers/finance/reports/agedDebtors/index.js:16]: Very well documented 👍

File: server/controllers/finance/reports/agedDebtors/index.js [hunk 5]
```diff
@@ -0,0 +1,209 @@
...
+const util          = require('../../../../lib/util');
+const ReportManager = require('../../../../lib/ReportManager');
+const BadRequest    = require('../../../../lib/errors/BadRequest');
+
+const TEMPLATE      = './server/controllers/finance/reports/agedDebtors/reports.handlebars';
+
+// expose to the API
+exports.document = document;
+
+/** processingAgedDebtorsReport */
```
Comment (mbayopanda) [inline server/controllers/finance/reports/agedDebtors/index.js:34]: Could you give the description of this function

File: server/controllers/finance/reports/agedDebtors/index.js [hunk 6]
```diff
@@ -0,0 +1,209 @@
...
+  return db.exec(query, [untilDate, untilDate]);
+}
+
+
+/**
+ * @function getFiscalYear
+ * @param {fiscalYear, untilDate} 
+ */
+function getPreviousDebts(fiscalYear, untilDate) {
+  let query =`
```
Comment (mbayopanda) [inline server/controllers/finance/reports/agedDebtors/index.js:147]: Could you give a description for this function

File: server/controllers/finance/trialBalance/index.js [hunk 7]
```diff
@@ -108,7 +108,7 @@ function checkDateInPeriod(transactions) {
   var sql =
     `SELECT COUNT(pj.uuid) AS count, pj.trans_id, pj.trans_date, p.start_date, p.end_date
     FROM posting_journal AS pj JOIN period as p ON pj.period_id = p.id
-    WHERE pj.trans_date NOT BETWEEN p.start_date AND p.end_date AND
+    WHERE DATE(pj.trans_date) NOT BETWEEN DATE(p.start_date) AND DATE(p.end_date) AND
```
Comment (mbayopanda) [inline server/controllers/finance/trialBalance/index.js:111]: 👍

**Review by mbayopanda — COMMENTED** [review]
The reports produced look very good, we just have to test it with big real data

**Comment by mbayopanda** [comment]
@lomamech could you update your branch and fix conflicts

**Comment by lomamech** [comment]
all things all Ok

@sfount @mbayopanda

**Comment by lomamech** [comment]
Improve the way to get information about the kind of repots 
Update interface of modal title

I will update in the next times, the E2E test for All Reports for checked the showing of the new report in the grid

@mbayopanda

File: server/controllers/finance/reports/agedDebtors/index.js [hunk 8]
```diff
@@ -0,0 +1,216 @@
...
+ * @function getFiscalYear
+ * @param {untilDate} 
+ */
+function getFiscalYear(untilDate) {
+  let query =`
+    SELECT fiscal_year.id, fiscal_year.previous_fiscal_year_id, period.id AS period_id, period.start_date, period.end_date
+    FROM fiscal_year
+    JOIN period ON period.fiscal_year_id = fiscal_year.id
+    WHERE period.start_date <= ? AND period.end_date >= ?;`;
+
```
Comment (mbayopanda) [inline server/controllers/finance/reports/agedDebtors/index.js:136]: Try to use DATE function to avoid errors with limits values

**Review by mbayopanda — CHANGES_REQUESTED** [review]
All things work well, excepts for extreme date `2016-12-31`. Could you try to fix it.

**Review by mbayopanda — APPROVED** [review]

**Comment by mbayopanda** [comment]
@lomamech could you update your branch. All things are good 👍

**Pull request merged**